wled00/button.cpp

```diff
@@ -131,21 +131,41 @@ void handleSwitch(uint8_t b)
   }
 }
 
+#define ANALOG_BTN_READ_CYCLE 250   // min time between two analog reading cycles
+#define STRIP_WAIT_TIME 6           // max wait time in case of strip.isUpdating() 
+#define POT_SMOOTHING 0.25f         // smoothing factor for raw potentiometer readings
+#define POT_SENSITIVITY 4           // changes below this amount are noise (POT scratching, or ADC noise)
+
 void handleAnalog(uint8_t b)
 {
-  static uint8_t oldRead[WLED_MAX_BUTTONS];
+  static uint8_t oldRead[WLED_MAX_BUTTONS]  = {0};
+  static float filteredReading[WLED_MAX_BUTTONS] = {0.0f};
+  uint16_t rawReading;    // raw value from analogRead, scaled to 12bit
+
   #ifdef ESP8266
-  uint16_t aRead = analogRead(A0) >> 2; // convert 10bit read to 8bit
+  rawReading = analogRead(A0) << 2;   // convert 10bit read to 12bit
   #else
-  uint16_t aRead = analogRead(btnPin[b]) >> 4; // convert 12bit read to 8bit
+  rawReading = analogRead(btnPin[b]); // collect at full 12bit resolution
   #endif
+  yield();                            // keep WiFi task running - analog read may take several millis on ESP8266
+
+  filteredReading[b] += POT_SMOOTHING * ((float(rawReading) / 16.0) - filteredReading[b]);  // filter raw input, and scale to [0..255]
+  uint16_t aRead = max(min(int(filteredReading[b]), 255), 0);                               // squash into 8bit
+  if(aRead <= POT_SENSITIVITY) aRead = 0;                                                   // make sure that 0 and 255 are used
+  if(aRead >= 255-POT_SENSITIVITY) aRead = 255;
 
   if (buttonType[b] == BTN_TYPE_ANALOG_INVERTED) aRead = 255 - aRead;
 
   // remove noise & reduce frequency of UI updates
-  aRead &= 0xFC;
+  if (abs(int(aRead) - int(oldRead[b])) <= POT_SENSITIVITY) return;  // no significant change in reading
+
+  // wait until strip finishes updating
+  unsigned long wait_started = millis();
+  while(strip.isUpdating() && (millis() - wait_started < STRIP_WAIT_TIME)) {
```

Comment:
This check is just added "to be safe", because I was not sure if the strip can start to update in parallel - Analogread can take a fee millis. If that's not possible, no problem I can remove the second check. I did not see flickering without the second wait.

Reply:
If you can test again I would be grateful since my current test set-up does not have potentiometer to validate.
In any case the code can be just commented out and if flickering (re)appears it can be uncommented quickly.

Reply:
I've tested without the second check. It work for me.
Actually there are other sources of "flickering", when changing effect speed or intensity. These are not related to analog input - when using web UI to change speed/intensity, there is also interruption in the effect.